care/emr/api/viewsets/valueset.py

```diff
@@ -91,3 +97,60 @@ def lookup_code(self, request, *args, **kwargs):
             .get()
         )
         return Response(result)
+
+    def _get_or_create_user_preferences(self, user) -> UserValueSetPreference:
+        try:
+            return UserValueSetPreference.objects.get(user=user)
+        except UserValueSetPreference.DoesNotExist:
+            with transaction.atomic():
+                return UserValueSetPreference.objects.get_or_create(
+                    user=user, defaults={"favorites": [], "recent_views": []}
+                )[0]
+
+    @action(detail=True, methods=["POST"])
+    def mark_favorite(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        preferences = self._get_or_create_user_preferences(request.user)
+        if len(preferences.get_favorites()) > preferences.MAX_FAVORITES:
+            raise ValidationError("Maximum number of favorites reached (50)")
+        preferences.add_favorite(valueset.external_id)
+        return Response({"message": "Marked favorite"})
+
+    @action(detail=True, methods=["POST"])
+    def remove_favorite(self, request, *args, **kwargs):
+        valueset = self.get_object()
+        preferences = self._get_or_create_user_preferences(request.user)
+        preferences.remove_favorite(valueset.external_id)
+        return Response({"message": "Removed from favorite"})
+
+    @action(detail=False, methods=["GET"])
+    def favorites(self, request):
+        preferences = self._get_or_create_user_preferences(request.user)
```

Comment:
instead of doing get or create, we could just return empty results as this is a get request